From 55dc46ac33f268a0bf878bcc6501cdcd1524696d Mon Sep 17 00:00:00 2001
From: Stefan Agner <stefan@agner.ch>
Date: Sun, 23 Jun 2019 14:53:59 +0200
Subject: [PATCH] desktop-shell: make sure child window stays active

If a xdg_toplevel surface has a child, the desktop shell still allows
to activate the parent. This can be problematic with modal dialogs
such as message boxes which then are hidden behind the main window,
which might be non-responsive to inputs at this point.

The protocol specifies set_parent as follows: "Set the 'parent' of
this surface. This surface should be stacked above the parent surface
and all other ancestor surfaces."

Track parent/child relationship in desktop-shell. Fllow the protocol
recommendation and make sure the child stays stacked above the parent.

Signed-off-by: Stefan Agner <stefan@agner.ch>
---
 desktop-shell/shell.c | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/desktop-shell/shell.c b/desktop-shell/shell.c
index 93b1c70b..ee5defd7 100644
--- a/desktop-shell/shell.c
+++ b/desktop-shell/shell.c
@@ -104,6 +104,9 @@ struct shell_surface {
 
 	struct desktop_shell *shell;
 
+	struct shell_surface *parent;
+	struct shell_surface *child;
+
 	int32_t saved_x, saved_y;
 	bool saved_position_valid;
 	bool saved_rotation_valid;
@@ -2433,6 +2436,9 @@ desktop_surface_removed(struct weston_desktop_surface *desktop_surface,
 	if (!shsurf)
 		return;
 
+	if (shsurf->parent)
+		shsurf->parent->child = NULL;
+
 	wl_signal_emit(&shsurf->destroy_signal, shsurf);
 
 	if (shsurf->fullscreen.black_view)
@@ -2747,6 +2753,20 @@ desktop_surface_resize(struct weston_desktop_surface *desktop_surface,
 		wl_resource_post_no_memory(resource);
 }
 
+static void
+desktop_surface_set_parent(struct weston_desktop_surface *desktop_surface,
+			   struct weston_desktop_surface *parent,
+			   void *shell)
+{
+	struct shell_surface *shsurf =
+		weston_desktop_surface_get_user_data(desktop_surface);
+	struct shell_surface *shsurf_parent =
+		weston_desktop_surface_get_user_data(parent);
+
+	shsurf->parent = shsurf_parent;
+	shsurf_parent->child = shsurf;
+}
+
 static void
 desktop_surface_fullscreen_requested(struct weston_desktop_surface *desktop_surface,
 				     bool fullscreen,
@@ -2929,6 +2949,7 @@ static const struct weston_desktop_api shell_desktop_api = {
 	.committed = desktop_surface_committed,
 	.move = desktop_surface_move,
 	.resize = desktop_surface_resize,
+	.set_parent = desktop_surface_set_parent,
 	.fullscreen_requested = desktop_surface_fullscreen_requested,
 	.maximized_requested = desktop_surface_maximized_requested,
 	.minimized_requested = desktop_surface_minimized_requested,
@@ -3792,6 +3813,12 @@ activate(struct desktop_shell *shell, struct weston_view *view,
 	shsurf = get_shell_surface(main_surface);
 	assert(shsurf);
 
+	/* Do not allow to activate surface which has a xdg child */
+	if (shsurf->child) {
+		activate(shell, shsurf->child->view, seat, flags);
+		return;
+	}
+
 	/* Only demote fullscreen surfaces on the output of activated shsurf.
 	 * Leave fullscreen surfaces on unrelated outputs alone. */
 	if (shsurf->output)
-- 
2.22.0

